Fake sync for Redshift tests (branch-conditional, refactored)

escherize opened the pull request:
Resolves DEV-1245

## Summary

Skip expensive Redshift sync during test setup by directly inserting Table/Field rows from test data definitions. After code review feedback from @camsaul, the implementation now:

1. **Runs real sync on master/release branches** - catches sync regressions in CI
2. **Uses fake sync on feature branches** - saves ~10 minutes per test run
3. **Separates transformation from insertion** - pure functions for testability

## The Journey

### Initial Implementation
- Added `use-fake-sync?` multimethod that always returned `true` for Redshift
- Directly inserted Table/Field rows from dbdef, skipping network calls

### Cam's Code Review Feedback
1. "I'm thinking we still want to make sure we don't break Redshift sync entirely" → Run real sync on master/release
2. "Separate the transformation from insertion for debuggability" → Refactored into pure + stateful functions
3. External table test was being skipped → Now runs on master/release where real sync happens

### Final Implementation
- Replaced `use-fake-sync?` multimethod with `:test/use-fake-sync` driver feature
- Added `on-master-or-release-branch?` to detect CI branch via `GITHUB_REF_NAME` or git
- Refactored `fake-sync-database!` into:
  - `field-def->row` - pure transformation
  - `table-def->rows` - pure transformation
  - `dbdef->fake-sync-rows` - pure, composes above
  - `insert-fake-sync-rows!` - stateful insertion
- Added `with-driver-supports-feature!` macro for testing

## Key Files

| File | Change |
|------|--------|
| `test/metabase/test/data/interface.clj` | Add `on-master-or-release-branch?`, `with-driver-supports-feature!`, `fake-sync-schema` |
| `test/metabase/test/data/impl/get_or_create.clj` | Refactored fake sync with pure/stateful separation |
| `modules/drivers/redshift/.../redshift.clj` | Branch-conditional `:test/use-fake-sync` |
| `src/metabase/driver.clj` | Add `:test/use-fake-sync` feature flag |

## Trade-offs

- No fingerprints (tests needing binning min/max must use real sync)
- External tables (Spectrum) only synced on master/release
- `database_type` simplified to base-type name

## Test Plan

- [x] `real-sync-validation-test` validates fake sync matches real sync
- [x] Native type maps handled correctly
- [x] External table test runs on master/release branches
- [x] Local Redshift test run completed successfully
- [x] Deterministic + pure tests for dataset definition -> tables + fields
- [ ] Full Redshift CI suite (runs on merge to master with real sync)

escherize committed:
fake sync for redshift

escherize committed:
Merge branch 'master' into fake-sync-redshift

escherize committed:
Add Ralph Wiggum loop files for fake-sync development

Set up autonomous development loop infrastructure:
- PROMPT_plan.md: Planning mode prompt with problem definition
- PROMPT_build.md: Build mode prompt for iterations
- IMPLEMENTATION_PLAN.md: Task checklist tracking progress
- WORK_LOG.md: Iteration observations and context dump
- AGENTS.md: Build/test commands
- loop.sh: Orchestration script

escherize committed:
add info about turning fake sync on and off

escherize committed:
Fix cyclic dependency in fake sync without requiring-resolve

Instead of using requiring-resolve to break the cycle, we untangle it
by simplifying fake sync to not need sql.tx at all:

1. Add tx/fake-sync-schema multimethod to interface.clj
   - Returns nil by default
   - Redshift implements it to return unique-session-schema

2. Simplify get_or_create.clj fake sync functions:
   - insert-fake-table!: Uses tx/fake-sync-schema (not sql.tx/session-schema)
   - insert-fake-field!: Uses (name base-type) for database_type
     (informational only - tests query by base_type)
   - insert-pk-field!: Hardcodes "id" (always "id" for SQL drivers)

This breaks the cycle cleanly without runtime resolution hacks.

escherize committed:
Enable fake sync for Redshift tests

Skip network sync calls by directly inserting Table/Field rows from test
data definitions. This reduces test setup time significantly by avoiding
RTT to AWS Redshift for metadata discovery.

Changes:
- Enable fake sync in `use-fake-sync?` multimethod
- Skip `test-external-table` when fake sync enabled (Spectrum external
  tables are not part of test dbdefs)

The fake sync implementation was added in commit a792314ce51.

escherize committed:
Fix fake sync to support custom and composite primary keys

The previous implementation always inserted an auto-generated "id" PK field,
which broke tests using custom PKs (e.g., `dataset-with-custom-composite-pk-test`).

Changes:
- Check if `(some :pk? field-definitions)` before adding auto PK
- Extract `:pk?` from FieldDefinition instead of passing as keyword arg
- Adjust field positions based on whether auto PK was inserted

This mirrors the logic in `sql/ddl.clj:add-pks-if-needed`.

escherize committed:
Add real-sync-validation-test and remove development files

- Add real-sync-validation-test that forces a real sync (via with-redefs)
  on a minimal dataset to validate fake sync doesn't diverge from real sync
- Remove Ralph Wiggum loop development files that were accidentally committed

escherize committed:
Fix fake sync to handle native type maps in base-type

escherize committed:
Remove development file

escherize committed:
fix indentation

escherize committed:
indentation 2

escherize committed:
loosen the mview-nm to find the mview-nm table

camsaul commented on test/metabase/test/data/impl/get_or_create.clj:
```diff
@@ -116,29 +116,153 @@
+                    (insert-auto-pk-field! driver table))
+                  ;; Insert fields with positions (0-indexed if custom PK, 1-indexed if auto PK)
+                  (dorun
+                   (map-indexed
```
I'm wondering if this might be a little more debuggable if you could separate out the code above into two discrete parts: 

1. Code to take dataset definition and return some representation of the Table and Field rows to insert into the app DB (as close as possible to the actual maps we'll pass to Toucan, but you'll need a placeholder for some of the IDs I guess)
2. Code to take the result of the previous step and actually insert it

Then the thread-unsafe `!` part of the code is kept to a minimum and you could easily test/debug the meat of the code doing the transformation from dataset definition to app DB rows

camsaul reviewed (approved):
Generally looks good but I think it would still be good to run redshift sync tests **sometimes**, maybe against `master` or even better as part of a nightly job (altho this would require a lot of work to set up)

escherize commented on test/metabase/test/data/interface.clj:
```diff
@@ -589,12 +620,12 @@
     :semantic_type :type/Quantity
     :name          "count"
     :display_name  (case aggregation-type
-                     :count     "Count"
```
missed one!

escherize committed:
address code review feedback

- External table test skipped with fake sync
  -> Fixed - now uses (when (tx/on-master-or-release-branch?) ...)

- Run real sync on master/release
  -> Implemented via on-master-or-release-branch?

- Separate transformation from insertion
  -> Done - pure fns (field-def->row, table-def->rows, dbdef->fake-sync-rows) + stateful fns (insert-fake-sync-rows!)

escherize committed:
indent 3

escherize committed:
improve naming (ddriver -> driver-arg, etc)

escherize committed:
Add pure transformation test for fake sync

Tests the dbdef->fake-sync-rows function that converts database definitions
to Table/Field row maps. This validates Cam's code review suggestion to
separate transformation logic from insertion - now we can test the "meat"
of the transformation without touching the database.

Covers:
- Auto PK injection at position 0
- User-defined field positioning
- FK semantic type assignment
- Native type maps ({:native ...} and {:natives ...} forms)

escherize committed:
Merge branch 'master' into fake-sync-redshift

escherize committed:
Run real fingerprinting after fake sync

Fake sync skips metadata discovery (tables/columns) but still needs
fingerprinting for tests that depend on min/max values, binning, etc.
This adds ~1-2 min but saves 4-8 min vs full sync.

escherize committed:
Add :type/DateTimeWithTZ mapping for Redshift test data

Tests using the `times-mixed` dataset (which has DateTimeWithTZ columns)
were failing with "datetimewithtz does not exist" because Redshift's
test data configuration was missing the type mapping.

Maps :type/DateTimeWithTZ to TIMESTAMPTZ, which is Redshift's
timezone-aware timestamp type.

escherize committed:
Pass scan level to fake-sync-database! for consistent behavior

Refactor to pass the scan level (:full or :schema) to fake-sync-database!
so fingerprinting is only run when scan is :full. This matches real sync
behavior where only test-data gets :full scan (with fingerprinting).

Fixes unix-timestamp-test which was failing because fingerprinting was
changing the effective_type of fields in custom datasets that shouldn't
be fingerprinted.

escherize committed:
Use sql.tx/field-base-type->sql-type for database_type in fake sync

Previously, fake sync used (name base-type) which returned the keyword
name like "DateTimeWithTZ". This caused SQL casts to use invalid types.

Now uses sql.tx/field-base-type->sql-type to get the proper SQL type
(e.g., "TIMESTAMPTZ" for Redshift's :type/DateTimeWithTZ).

escherize committed:
Fix unix-timestamp-test to work with fake sync

Use merged-mock-metadata-provider pattern instead of with-temp-vals-in-db
for setting coercion strategy. The with-temp-vals-in-db approach can have
caching issues with metadata providers, while merged-mock-metadata-provider
directly wraps the provider with the overridden field attributes.

This pattern is already used in coercion_test.clj for similar tests.

Also added :effective-type :type/Decimal to the dataset definition since
fake sync needs explicit effective types for native column types.

escherize committed:
Merge branch 'master' into fake-sync-redshift

escherize committed:
Fix fake-sync-transformation-test assertion

When effective-type is provided in a native type field definition,
base_type should use that value, not :type/*. The :type/* fallback
only applies when no effective-type is specified.
